### ISSUE TITLE
twisted.test.test_tcp.AddressTestCase.testBuildProtocol is buggy and overly complex

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2831]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2831|
|Type|defect|
|Created|2007-09-12 02:16:35Z|

This test method seems to assert that ServerFactory.buildProtocol and ClientFactory.buildProtocol are each called with an address object which correctly reflects what has happened on the network. It is missing documentation to this effect, though.

It does much more work in order to set up state which can be asserted against than is necessary - there's no reason to pre-bind and then close a socket; the bind address of the client can be discovered after the fact, rather than specified in advance.

The test is somewhat non-deterministic in that it is really unspecified whether the makeConnection method of the server or the client side of the connection will run first.  Realistically, this will always go one way or the other for a particular reactor/platform combination.  However, it is sloppy to track state the way this test does, setting a flag which indicates success from two different, unrelated codepaths.

The factoring is also rather poor, as the asserts happen outside of the test method itself in callbacks which happen to be in the codepath being tested.

Lastly, it contains a race condition which will cause it to sometimes complete while one side or the other of the connection is still connected, probably while still having an outstanding DelayedCall to disconnect, triggering a warning from trial.

<details><summary>Searchable metadata</summary>

```
trac-id__2831 2831
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1189563395000000 1189563395000000
changetime__1189607504000000 1189607504000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

I wrote a new version of the test in test-build-protocol-2831.  It didn't end up as much shorter than the original than I hoped, but I guess that's at least partly because it has docstrings and some comments (and it's still a little shorter).  But I didn't actually say anything about the code being too long in the ticket description, so I guess that's okay. :p

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @exarkun|
|-|-|

 * ```MyServerFactory``` and ```MyClientFactory``` don't have docstrings, especially for the attributes you added
 * There are some unclear areas around ```RememberingWrapper``` being a ```ClientFactory```, but not forwarding calls to ```clientConnectionLost```/```clientConnectionFailed```, and being used for wrapping the server too (the latter isn't much an issue, just a small comment would clarify this). 
 * I think ```MyClientFactory.deferred``` and ```MyClientFactory.protocolConnectionLost``` are more or less redundant.
 * test_buildProtocol should also test the output of the getPeer method, as it was done before.
 * cbDisconnected should check ```clientAddr[0].host``` and ```serverAddr[0].host``` instead of 127.0.0.1

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

Thanks for a great review :)

I factored most of the new stuff and some old stuff out of MyServerFactory and MyClientFactory into a mixin that they now both share, it has docs for some old attributes and the new one.

I added a couple comments to RemeberingWrapper, let me know if I should try to make it more clear.

Hmm yea, I didn't catch the meaning of MyClientFactory.deferred before.  I'd rather leave it alone for now, if that's okay with you.  The test needs that functionality from both the client and server factories, and MyServerFactory didn't have anything like it :/

Okay, I split test_buildProtocol into four chunks: set up, tear down, and one test method for the client side and the server side.  I added appropriate assertions for getPeer to each test method.  setUp is still kind of long by itself, but the rest almost seem reasonable, even with the extra assertions.

cbDisconnected used 127.0.0.1 since on windows you can't connect to 0.0.0.0, which is what the host attribute will say -- unless you pass 127.0.0.1 to listenTCP for the interface argument, so I did that, and changed the tests to use the attribute instead of the literal.

Doing a build on buildbot now, but I think everything should still work.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

buildbot is green, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [21278]) Merge test-build-protocol-2831

Author: exarkun
Reviewer: therve
Fixes [#2831](https://github.com/twisted/twisted/issues/2831)

Rewrite the old test case which covered the value passed to buildProtocol as
the remote address and the values returned by the getPeer and getHost methods
of tcp transports as two new tests which cover the same behavior but do not
have a race condition in their cleanup.  This removes a very intermittent
failure which would leave a DelayedCall and a tcp transport in the reactor
causing trial to report a warning.  The new tests are also more thoroughly
documented and should be easier to maintainer.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
